### PR TITLE
Java: CWE-552 Query to detect unsafe resource loading in Java Spring applications

### PR DESCRIPTION
**Pull request by luchua-bc**
Directly incorporating user input into loading resource calls on the server side without proper validation of the input can allow any web application resource such as configuration files and source code to be disclosed.

It is very common that Java Spring applications load requested resources and return file contents to clients. Constructing a server-side URI path with user input could allow an attacker to download application binaries
(including application classes or jar files) or view arbitrary files within protected directories.
 
This query detects unsafe usage of resource loading in Spring including the following APIs:
- ClassPathResource
- ResourceUtils
- ResourceLoader and ApplicationContext

It utilizes the path sanitizer library to reduce FPs:
- Path traversal check
- Path encoding check
- Check of path normalization using the java.nio.file.Path package

Please consider to merge the PR. Thanks.

### CODE REVIEW

**Review by atorralba — COMMENTED** [review]
Hey @luchua-bc, thanks for your contribution. I added a couple comments that should be addressed if you're still interested in merging this PR regardless of bounty status.

**Comment by luchua-bc** [comment]
Hi @atorralba,

I think this query will be beneficial so I'd like to have it merged. I've made requested changes.

@luchua-bc

File: java/ql/test/experimental/query-tests/security/CWE-552/UnsafeLoadSpringResource.java [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+				}
+				content = out.toString();
+			} catch (IOException ie) {
+				ie.printStackTrace();
+			}
+		}
+		return content;
+	}
+
+	/**
```
[the remark this answers is not in the document]
Comment (luchua-bc) [reply java/ql/test/experimental/query-tests/security/CWE-552/UnsafeLoadSpringResource.java]: I've removed the code from the relevant CVE and added simplified code.

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 2]
```diff
@@ -175,3 +192,17 @@ private class FilePathFlowStep extends SummaryModelCsv {
...
 }
+
+/** Taint model related to resource loading in Spring. */
+private class LoadSpringResourceFlowStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.springframework.core.io;ClassPathResource;false;ClassPathResource;;;Argument[0];Argument[-1];taint",
+        "org.springframework.core.io;ClassPathResource;true;" +
+          ["getFilename", "getPath", "getURL", "resolveURL"] + ";;;Argument[-1];ReturnValue;value",
```
[the remark this answers is not in the document]
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Changed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 3]
```diff
@@ -175,3 +192,17 @@ private class FilePathFlowStep extends SummaryModelCsv {
...
   }
 }
+
+/** Taint model related to resource loading in Spring. */
+private class LoadSpringResourceFlowStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.springframework.core.io;ClassPathResource;false;ClassPathResource;;;Argument[0];Argument[-1];taint",
+        "org.springframework.core.io;ClassPathResource;true;" +
```
[the remark this answers is not in the document]
Comment (atorralba) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:202]: Hm, this means tests for these steps are also needed @luchua-bc.

[the remark this answers is not in the document]
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:202]: I've appended `;manual` to all the rows.

@atorralba - sorry I didn't get what you mean for `tests for these steps are also needed`. I already created test cases for `ClassPathResource`, `ResourceLoader`, and `Resource`. Please clarify.

[the remark this answers is not in the document]
Comment (atorralba) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:202]: I mean tests should be expanded to include these taint steps. So, test that `ClassPathResource.(getFilename|getPath|getURL|resolveURL)`, `ResourceLoader.getResource`, and `Resource.createRelative` propagate taint as expected before reaching a sink.

**Comment by pwntester** [comment]
Hi! after reviewing the LGTM run results, I spot a number of FPs caused by:

- Non-classpath related `InputStreamSource`s such as [`MultipartFile.getResource`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/multipart/MultipartFile.html#getResource--) should not be considered as sinks. 
  - sample project exhibiting these FPs: BanqiJane/Bilibili_Danmuji
- [`HttpEntity.getBody`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/HttpEntity.html#getBody--). Not sure why this was reported or considered a sink in the first place but should be excluded
  - sample project exhibiting these FPs: RomanKonovalov/securityalarm 
- [ResourceUtils.toURI](https://lgtm.com/projects/g/spring-projects/spring-framework/snapshot/a3295913623382f55768e84c5d38a2537840fc6f/files/spring-core/src/main/java/org/springframework/util/ResourceUtils.java#xa8e7cf3b922f6149) not sure why this is considered a sink, if any, it should be considered a taint step, doesnt it?
  - sample project exhibiting these FPs: spring-projects/spring-framework 
- org.springframework.core.io.Resource interface should not be used, we should probably use its implementing [`ClasspathResource`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/io/ClassPathResource.html) instead.
  - sample project exhibiting these FPs: HongZhaoHua/jstarcraft-core

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:210]: I've added test case for the first two categories - `ClassPathResource.getFilename` and `ResourceLoader.getResource`. `Resource` related ones are no longer applicable as per discussion below.

**Comment by luchua-bc** [comment]
Hi @pwntester,

As per your suggestion, I've removed `Resource` and `InputStreamSource` from the model to reduce FPs. FPs from those 4 mentioned repositories disappeared after the change.

@luchua-bc

**Comment by pwntester** [comment]
> Hi @pwntester,
> 
> As per your suggestion, I've removed `Resource` and `InputStreamSource` from the model to reduce FPs. FPs from those 4 mentioned repositories disappeared after the change.
> 
> @luchua-bc

Thanks a lot!

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 4]
```diff
@@ -175,3 +188,33 @@ private class FilePathFlowStep extends SummaryModelCsv {
...
+      [
+        // Get spring resource
+        "org.springframework.core.io;ClassPathResource;true;" +
+          ["getFilename", "getPath", "getURL", "resolveURL"] + ";;;Argument[-1];get-resource;manual",
+        // "org.springframework.core.io;Resource;true;" +
+        //   ["getFile", "getFilename", "getURI", "getURL"] +
+        //   ";;;Argument[-1];get-resource;manual",
+        // "org.springframework.core.io;InputStreamSource;true;" +
+        //   ["getInputStream"] +
+        //   ";;;Argument[-1];get-resource;manual"
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:217]:
```suggestion
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 5]
```diff
@@ -99,6 +102,16 @@ private class GetResourceSink extends UnsafeUrlForwardSink {
   }
 }
 
+/** Sink of Spring resource loading. */
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:105]:
```suggestion
/** A sink for methods that load Spring resources. */
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 6]
```diff
@@ -175,3 +188,33 @@ private class FilePathFlowStep extends SummaryModelCsv {
       ]
   }
 }
+
+/** Taint model related to resource loading in Spring. */
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:192]:
```suggestion
/** Taint models related to resource loading in Spring. */
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 7]
```diff
@@ -175,3 +188,33 @@ private class FilePathFlowStep extends SummaryModelCsv {
...
+    row =
+      [
+        "org.springframework.core.io;ClassPathResource;false;ClassPathResource;;;Argument[0];Argument[-1];taint;manual",
+        "org.springframework.core.io;ResourceLoader;true;getResource;;;Argument[0];ReturnValue;taint;manual",
+        "org.springframework.core.io;Resource;true;createRelative;;;Argument[0];ReturnValue;taint;manual"
+      ]
+  }
+}
+
+/** Sink related to spring resource. */
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:204]:
```suggestion
/** Sink models for methods that load Spring resources. */
```

File: java/ql/src/experimental/semmle/code/java/frameworks/SpringResource.qll [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+private import semmle.code.java.dataflow.ExternalFlow
+private import semmle.code.java.dataflow.FlowSources
+
+/** A utility class for resolving resource locations to files in the file system in the Spring framework. */
+class ResourceUtils extends Class {
+  ResourceUtils() { this.hasQualifiedName("org.springframework.util", "ResourceUtils") }
+}
+
+/**
+ * Resource loading method declared in Spring `ResourceUtils`.
```
Comment (atorralba) [inline java/ql/src/experimental/semmle/code/java/frameworks/SpringResource.qll:15]:
```suggestion
 * A method declared in `org.springframework.util.ResourceUtils` that loads Spring resources.
```

**Review by atorralba — COMMENTED** [review]
Hey @luchua-bc sorry that this stalled for so long. I think if we remove the commented code and address the QLDoc style warnings above, this should be ready to be merged.

**Comment by luchua-bc** [comment]
Hi @atorralba,

No problem and I know you are busy.

Those text changes are pretty simple to make but it seems that there were other changes in that directory and in the main branch that impacts this query. When I run the query, I got some warnings like:
```
WARNING: Predicate isSanitizerGuard overrides only deprecated predicates, but is not itself deprecated (/github/vscode-codeql-starter/ql/java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql:40,3-42,4)
WARNING: Type BarrierGuard has been deprecated and may be removed in future (/github/vscode-codeql-starter/ql/java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql:40,39-61)
WARNING: Type HttpServletRequestGetRequestURIMethod has been deprecated and may be removed in future (/github/vscode-codeql-starter/ql/java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql:28,22-59)
WARNING: Type PathTraversalBarrierGuard has been deprecated and may be removed in future (/github/vscode-codeql-starter/ql/java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql:41,22-47)
``` 

And the expected query results don't match actual results. I think I need to synchronize with the main branch then add changes on top of it. I will further investigate and get it done before next Monday.

@luchua-bc

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 5 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Updated

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 6 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Updated

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 7 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Updated

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 4 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Removed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 9]
```diff
@@ -175,3 +188,27 @@
...
+
+/** Sink models for methods that load Spring resources. */
+private class SpringResourceCsvSink extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        // Get spring resource
+        "org.springframework.core.io;ClassPathResource;true;" +
+          ["getFilename", "getPath", "getURL", "resolveURL"] + ";;;Argument[-1];get-resource;manual",
+      ]
```
[the remark this answers is not in the document]
Comment (atorralba) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:212]:
```suggestion
      // Get spring resource
      "org.springframework.core.io;ClassPathResource;true;" +
        ["getFilename", "getPath", "getURL", "resolveURL"] + ";;;Argument[-1];get-resource;manual"
```

**Comment by luchua-bc** [comment]
Thank @atorralba for committing the change.

**Review by atorralba — APPROVED** [review]

**Pull request merged**